### DISCUSSION TITLE
SVG images in memory (no read/write (temporary files))

### DISCUSSION THREAD

**Discussion by DocMinus** [rdkit/rdkit#5996] [Show and tell]
Perhaps this is obvious to some, but I couldn't really find any examples dealing with displaying (SVG image) structures that don't go via using (temporary files). I am talking in context of e.g. using a GUI. (edit) Example  would be when you don't have ipython console available, say when you use tkinter or you use as webpage. 

I have put up a GIST (see below), if anyone is interested. The example is of course simple, and in Jupyter one would just use the rdkit method as is also shown, but it proves the concept.
Myself I have used this together with plotly dash to display structures without having to write temporary files (shown in the comment section of the GIST).
[https://gist.github.com/DocMinus/026ed8c845f9ad3c7dcf3df286038714](https://gist.github.com/DocMinus/026ed8c845f9ad3c7dcf3df286038714)

Again, it feels in retrospect simple/obvious, but I haven't encountered any solutions close to this so far. Thanks for reading.
Edit: fixed the link

**Comment by ptosco** [comment]
@DocMinus The GIST link does not work. You can generate a base64-encoded `<img>` HTML tag as follows:
```
from rdkit import Chem
from rdkit.Chem.Draw import rdMolDraw2D
from IPython.display import SVG

drawer = rdMolDraw2D.MolDraw2DSVG(-1, -1)
drawer.DrawMolecule(Chem.MolFromSmiles('N1CCCCC1'))
drawer.FinishDrawing()
svg = drawer.GetDrawingText()
svg_base64 = base64.b64encode(svg.encode("utf-8")).decode("utf-8")
HTML(f'<img src="data:image/svg+xml;base64,{svg_base64}">')
```
![image](https://user-images.githubusercontent.com/5244385/213751578-20c864d6-d3ea-44e6-ba89-65bda5d366cd.png)

**Reply by DocMinus** [reply]
@ptosco I fixed the link. The whole idea is when you don't have ipython console available, say when you use tkinter or you use as webpage. otherwise you obviously have the above solution (also shown in my GIST). 
I can maybe clarify my post above. Thanks for answering though!

**Comment by ptosco** [comment]
I have updated my post above to use an HTML tag. I am not sure what you are attempting to do and why you think an intermediate file is needed. It is worth mentioning that you can always generate file-like object in memory using `io.StringIO` (text files) and `io.ByesIO` (binary files) that you can read from and write to just like you would do with a temporary file on disk. Happy to help further if you wish to clarify what is your goal.

**Reply by DocMinus** [reply]
Thanks. Well, that is basically what I had :D 
My post wasn't so much a question than a sharing. But it wasn't in vain, should you get that impression - the io.StringIO was new to me. Will play around with that when I get a chance.
